stcrpy/tcr_metrics/tcr_dockq.py

```diff
@@ -0,0 +1,405 @@
+
+        This method evaluates the quality of a predicted TCR-pMHC complex (the "dock" structure)
+        against a reference (native) structure using the DockQ scoring system. It supports both
+        merged TCR-pMHC interfaces and separate chain evaluations, depending on the class setting.
+
+        Args:
+            dock (abTCR): The predicted TCR-pMHC complex structure to be evaluated.
+            reference (abTCR): The reference (native) TCR-pMHC complex structure.
+            save_merged_complex (bool, optional): If True, saves the merged complex structures
+                as PDB files for further inspection. Defaults to False.
+
+        Returns:
+            dict: A dictionary containing DockQ results, including DockQ score, F1, iRMSD, LRMSD,
+                fnat, fnonnat, and other relevant metrics for the best mapping.
+        """
+
+        try:
+            import DockQ
+            from DockQ import DockQ as dockq
+        except (ImportError, ModuleNotFoundError) as e:
+            warnings.warn(
+                f"DockQ is not installed. Error: {e}. Please install it using `pip install DockQ`."
+            )
+
+        mapping, string_mapping = self.try_tcr_mapping(dock, reference)
+
+        filtered_dock, filtered_reference = self.filter_residues(
+            dock, reference, mapping
+        )
+
+        model_structure = self.tcr_to_structure(
+            filtered_dock,
+            mapping_filter=mapping[0],
+            merge_chains=self.TCR_pMHC_interface,
+        )
+        native_structure = self.tcr_to_structure(
+            filtered_reference,
+            mapping_filter=mapping[1],
+            merge_chains=self.TCR_pMHC_interface,
+        )
+
+        # save the structures to pdb files
+        import Bio
+
+        io = Bio.PDB.PDBIO()
+        io.set_structure(model_structure)
+        io.save(f"model_structure_{dock.parent.parent.id}_{dock.id}.pdb")
+        io.set_structure(native_structure)
+        io.save(f"native_structure_{reference.parent.parent.id}_{reference.id}.pdb")
+
+        if self.TCR_pMHC_interface:
+            string_mapping = "TM:TM"
+
+        initial_mapping, model_chains, native_chains = dockq.format_mapping(
+            string_mapping, small_molecule=self.args.small_molecule
+        )
+
+        model_structure = dockq.load_PDB(
+            f"model_structure_{dock.parent.parent.id}_{dock.id}.pdb",
+            chains=model_chains,
+            small_molecule=self.args.small_molecule,
+        )
+        native_structure = dockq.load_PDB(
+            f"native_structure_{reference.parent.parent.id}_{reference.id}.pdb",
+            chains=native_chains,
+            small_molecule=self.args.small_molecule,
+        )
+
+        if not save_merged_complex:
+            os.remove(f"model_structure_{dock.parent.parent.id}_{dock.id}.pdb")
+            os.remove(f"native_structure_{reference.parent.parent.id}_{reference.id}.pdb")
+
```

Comment:
I sometimes like to use tempfile.tempdir for this type of thing so that it can all be done in a context manager and the ghost files don't apply in the working directory and also it will always clean up the files even if something errors in the middle.

Reply:
+1